### PR TITLE
#24303 ActorSystem.getWhenTerminated which returns CompletionStage

### PR DESCRIPTION
**Pull request by evis**

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

File: akka-actor-tests/src/test/java/akka/actor/ActorSystemTest.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
+package akka.actor;
```
Comment (ktoso) [inline akka-actor-tests/src/test/java/akka/actor/ActorSystemTest.java]: Needs copyright header, I'll add it.

**Review by ktoso — APPROVED** [review]
LGTM, nice thanks!

**Comment by johanandren** [comment]
Needs binary compatibility excludes as it introduces new methods. (see more here: https://github.com/akka/akka/blob/master/CONTRIBUTING.md#binary-compatibility )

**Comment by evis** [comment]
I guess, we need to restart PR validation once again? Seems like validation failed due to some `git fetch` fail.

**Review by patriknw — APPROVED** [review]
LGTM

**Comment by johanandren** [comment]
PLS BUILD

**Comment by johanandren** [comment]
Fixes #24303

File: akka-actor-typed-tests/src/test/java/akka/actor/typed/ActorSystemTest.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    final CompletionStage<Terminated> cs = system.getWhenTerminated();
+    cs.toCompletableFuture().get(2, SECONDS);
+  }
+
+  @Test
+  public void testGetWhenTerminatedWithoutTermination() throws Exception {
+    final ActorSystem system = ActorSystem.create(Behavior.empty(), "GetWhenTerminatedWithoutTermination");
+    final CompletionStage<Terminated> cs = system.getWhenTerminated();
+    try {
+      cs.toCompletableFuture().get(2, SECONDS);
```
Comment (johanandren) [inline akka-actor-typed-tests/src/test/java/akka/actor/typed/ActorSystemTest.java]: I think this test should rather just look at `isDone` and assert it is false than sleep 2 seconds every time this test is run.

**Review by johanandren — APPROVED** [review]
Impl LGTM but would be better if test didn't always sleep two seconds.

File: akka-actor-typed-tests/src/test/java/akka/actor/typed/ActorSystemTest.java (hunk 2 above)
Comment (evis) [reply akka-actor-typed-tests/src/test/java/akka/actor/typed/ActorSystemTest.java]: What would it test, though? We could just alright remove such test then... I mean, such test succeeds:

```
@Test
public void test() throws Exception {
  system.terminate();
  final CompletionStage<Terminated> cs = system.getWhenTerminated();
  Assert.assertFalse(cs.toCompletableFuture().isDone());
}
```

So, `isDone` doesn't seem to be the right way to check, if system will be terminated soon.

Comment (johanandren) [reply akka-actor-typed-tests/src/test/java/akka/actor/typed/ActorSystemTest.java]: Why would it be terminated soon if you didn't terminate it, and if you need to test that it doesn't become terminated on its own why is 2 seconds enough?

I think the only value of the current test is to verify that when the actor system is not terminated, the CS isn't completed either, and that can be tested just fine with `isDone`. 

The other test should still do the blocking wait for termination though, as it verifies that some time soon after terminating the actor system the CS completes.

Comment (evis) [reply akka-actor-typed-tests/src/test/java/akka/actor/typed/ActorSystemTest.java]: But when the actor system is terminated, the CS isn't completed immediately too, what is shown in the test from my previous comment :)

I suggest to just decrease timeout from 2 seconds to something like 200 millis, if 2 seconds is unacceptable. Then, if some future change will broke this method, test will fail at least sometimes.

Comment (johanandren) [reply akka-actor-typed-tests/src/test/java/akka/actor/typed/ActorSystemTest.java]: What you showed in the sample in the comment is a indeterministic test, you can never know that the CS completion has happened before that assertion is reached or not, so for that case it makes sense to wait for a maximum time period, when we know the completion will arrive.

To make the opposite test, that verifies that the CS doesn't magically complete on its own, you would have to wait forever though, which isn't really very useful.

What is potentially useful is to know that the CS returned from a running system isn't already completed, if we poll this during an hour, two seconds, 200 milliseconds or just check it right away doesn't really matter - we wont be more certain that it doesn't magically complete just after our span ends. Therefore `isDone` is perfectly fine for this test.

Comment (evis) [reply akka-actor-typed-tests/src/test/java/akka/actor/typed/ActorSystemTest.java]: Ok, thanks for detailed explanation. Now that makes sense to me too.

**Review by johanandren — APPROVED** [review]
LGTM, thanks!

**Comment by raboof** [comment]
Refs #24303

**Pull request merged**